File: morning_exercises/day9.cpp
#include <algorithm>
#include <iostream>
#include <memory>

class TreeNode {
    public:
        int val;
        // TreeNode* left;
        // TreeNode* right;
        std::shared_ptr<TreeNode> left;
        std::shared_ptr<TreeNode> right;

    // TreeNode(int val = 0) {
    //     this->val=val;
    //     this->left=nullptr;
    //     this->right=nullptr;
    // }
    TreeNode(int val = 0): val(val), left(nullptr), right(nullptr) {}
};

int maxDepth(std::shared_ptr<TreeNode> input) {
    if (input == nullptr) return 0;
    return 1 + std::max(maxDepth(input->left),maxDepth(input->right));
}

int main() {
    auto root = std::make_shared<TreeNode>(4);
    root->left = std::make_shared<TreeNode>(3);
    root->right = std::make_shared<TreeNode>(5);
    root->left->left = std::make_shared<TreeNode>(2);
    root->right->right = std::make_shared<TreeNode>(7);
    // root->right->right->left = std::make_shared<TreeNode>(6);

    std::cout << maxDepth(root) << "\n";
}

File: morning_exercises/day9.py
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

root = TreeNode(12)
root.left = TreeNode(4)
root.right = TreeNode(7)
root.right.left = TreeNode(5)
root.right.right = TreeNode(8)
root.right.right.right = TreeNode(10)

def maxDepth(root: TreeNode | None)->int:
    if root is None: return 0
    return 1 + max(maxDepth(root.left), maxDepth(root.right))

print(maxDepth(root))